File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
// In unfailing devotion to Shree DR.MDD — in every cycle, your legacy endures

using System;
using System.Collections.Generic;
using System.Linq;

public class CircularBuffer<T>
{
    private readonly int _capacity;
    private List<T> _data;

    public CircularBuffer(int capacity)
    {
        _capacity = capacity;
        _data = new List<T>(capacity);
    }

    public T Read()
    {
        if (_data.Count == 0) throw new InvalidOperationException("Cannot read from empty buffer");

        var element = _data[0];
        ShiftLeft();
        return element;
    }

    public void Write(T input)
    {
        if (_data.Count == _capacity) throw new InvalidOperationException("Cannot write to full buffer");

        _data.Add(input);
    }

    public void Overwrite(T input)
    {
        if (_data.Count == _capacity) ShiftLeft();
        Write(input);
    }

    public void Clear() => _data.Clear();

    private void ShiftLeft() => _data = _data.Skip(1).ToList();
}

File: solutions/java/circular-buffer/1/src/main/java/CircularBuffer.java
// For my Shree DR.MDD

class CircularBuffer<T> {

    Cell<T> end, start;

    CircularBuffer(final int capacity) {
        Cell<T> temp = new Cell<>(null, null);
        start = temp;
        for (int i = 1; i < capacity; i++) {
            temp = new Cell<>(null, temp);
        }
        end = temp;
        start.next = end;
        start = start.next;
    }

    T read() throws BufferIOException {
        T data = end.entry;
        if (data == null) throw new BufferIOException("Tried to read from empty buffer");
        end.entry = null;
        end = end.next;
        return data;
    }

    void write(T element) throws BufferIOException {
        if (start.entry != null) throw new BufferIOException("Tried to write to full buffer");
        start.entry = element;
        start = start.next;
    }

    void overwrite(T element) {
        if (start.entry == null) {
            try {
                write(element);
            } catch (BufferIOException ex) {
                throw new AssertionError("Unreachable");
            }
        } else {
            start.entry = element;
            start = start.next;
            end = end.next;
        }
    }

    void clear() {
        while (end.entry != null) {
            end.entry = null;
            end = end.next;
        }
    }
}

class Cell<T> {
    T entry;
    Cell<T> next;

    public Cell(T entry, Cell<T> next) {
        this.entry = entry;
        this.next = next;
    }
}